File: low-level-design/design-patterns/python/behavioural/state-pattern/suspended_state.py
from abc import ABC,abstractmethod





class SuspendedError(Exception):
    pass 

from state import VMState


class SuspendedState(VMState):
    def start(self, vm):
        from state_factory import StateFactory
        print(f"Starting VM {vm.name}")
        vm.set_state(StateFactory.get_starting_state())
    
    def stop(self, vm):
        print(f"Cannot stop {vm.name} as it is already suspended")
        return SuspendedError("Cannot stop VM as it is already suspended")
    
    def restart(self, vm):
        from state_factory import StateFactory
        print(f"Restarting VM {vm.name}")
        vm.set_state(StateFactory.get_restarting_state())
    
    def suspend(self, vm):
        print(f"VM {vm.name} is already in suspended state, no action performed")
        return SuspendedError("VM is already in suspended state, no action performed")
    
    def get_name(self):
        return "Suspended"

File: low-level-design/design-patterns/golang/behavioural/state-pattern/state/suspended_state.go
package state

import (
	"errors"
	"fmt"
)

func NewSuspendedState() *SuspendedState {
	return &SuspendedState{}
}

type SuspendedState struct {
}

func (s *SuspendedState) Start(vm *VM) error {
	fmt.Printf(" starting the vm  %s \n", vm.name)
	vm.SetState(NewStartingState())
	return nil
}
func (s *SuspendedState) Stop(vm *VM) error {
	fmt.Printf(" stopping the vm  %s \n", vm.name)
	vm.SetState(NewStoppedState())
	return nil
}

func (s *SuspendedState) Restart(vm *VM) error {
	fmt.Printf(" restarting the vm  %s \n", vm.name)
	vm.SetState(NewRestartingState())
	return nil
}
func (s *SuspendedState) Suspend(vm *VM) error {
	fmt.Printf("%s vm is already suspended \n", vm.name)
	return errors.New("vm is already suspended")

}

func (s *SuspendedState) GetName() string {
	return "Suspended"
}
